syft/execution/plan.py

```diff
@@ -246,17 +247,19 @@ def add_placeholder(self, tensor, node_type=None):
 
             self.var_count += 1
 
-        return self.placeholders[tensor.id]
+        return ObjectId(tensor.id)
 
-    def replace_with_placeholders(self, obj, **kw):
+    def replace_with_placeholder_ids(self, obj, **kw):
         """
-        Replace in an object all FrameworkTensors with Placeholders
+        Replace in an object all FrameworkTensors with Placeholder ids
         """
         if isinstance(obj, (tuple, list)):
-            r = [self.replace_with_placeholders(o, **kw) for o in obj]
+            r = [self.replace_with_placeholder_ids(o, **kw) for o in obj]
```

Comment:
I'm not sure I understand the strategy: you're now sotring ObjectIds instead of tags right? What are the benefits?

Reply:
You remember the `self.worker._tmp_placeholders` variable that held placeholders during deserialization in order to avoid creating multiple identical `Placeholders`, because duplicate `Placeholders` breaks instantiation with actual tensors? This is a strategy for getting rid of that, by storing `Placeholders` at the `Plan` level and using that store to avoid duplication.

Reply:
Using ids here makes it possible to deserialize without creating duplicate `Placeholders` and then index into the `Placeholder` store on the `Plan` to make sure that all the places that reference the same id use the same `Placeholder` object.